TrkDiag/src/TrackResolution_module.cc

```diff
@@ -0,0 +1,166 @@
+    };
+
+    using Parameters = art::EDProducer::Table<Config>;
+    TrackResolution(const Parameters& conf);
+
+  private:
+    void produce(art::Event& event) override;
+    int resolve(const KalSeed* k_1, const KalSeed* k_2);
+
+    art::InputTag _seedTag;
+    bool _useCaloHit;
+    bool _makeHists;
+    int _debug;
+
+    enum {kFirst, kSecond, kNeither}; // possible resolution results
+
+    struct Hist_t {
+      TH1* fitcon;
+    };
+    Hist_t _hists;
+  };
+
+  //--------------------------------------------------------------------------------------------------------
+  TrackResolution::TrackResolution(const Parameters& conf) :
+    art::EDProducer{conf},
+    _seedTag(conf().seedTag()),
+    _useCaloHit(conf().useCaloHit()),
+    _makeHists(conf().makeHists()),
+    _debug(conf().debug())
+  {
+    produces<KalSeedPtrCollection>();
+    if(_makeHists) {
+      art::ServiceHandle<art::TFileService> tfs;
+      art::TFileDirectory tfdir = tfs->mkdir("resolving");
+      _hists.fitcon = tfdir.make<TH1F>("fitcon", "Track p(#chi^2)", 200,  0.,  1.);
+    }
+  }
+
+  //--------------------------------------------------------------------------------------------------------
+  // Decide which track is better
+  int TrackResolution::resolve(const KalSeed* k_1, const KalSeed* k_2) {
+    if(!k_1 && !k_2) return kNeither;
+    if(!k_1) return kSecond;
+    if(!k_2) return kFirst;
+
+    const float fitcon_1(k_1->fitConsistency()), fitcon_2(k_2->fitConsistency());
+    if(_debug > 1) printf("    Comparing two tracks: Calo cluster: k1 = %o, k2 = %o; Fit quality: k1 = %.3g, k2 = %.3g\n",
+                          k_1->hasCaloCluster(), k_2->hasCaloCluster(), fitcon_1, fitcon_2);
+
+    // If requested, prioritize tracks that include a calorimeter cluster
+    if(_useCaloHit ) {
+      if( k_1->hasCaloCluster() && !k_2->hasCaloCluster()) return kFirst;
+      if(!k_1->hasCaloCluster() &&  k_2->hasCaloCluster()) return kSecond;
+    }
+
+    // Compare the fit quality
+    if(fitcon_1 > fitcon_2) return kFirst;
+    return kSecond;
+  }
+
+  //--------------------------------------------------------------------------------------------------------
+  void TrackResolution::produce(art::Event& event ) {
+    // create output
+    std::unique_ptr<KalSeedPtrCollection> trkcol(new KalSeedPtrCollection());
+
+    // get the KalSeedPtrs
+    auto handle = event.getValidHandle<KalSeedPtrCollections>(_seedTag);
+    const auto& clusters = *handle;
+    if(_debug > 0) printf("[TrackResolution::%s::%s] Collection of %zu track clusters retrieved\n", __func__, moduleDescription().moduleLabel().c_str(), clusters.size());
+
+    // Go through the tracks and check for alternate track fit hypotheses using the same helix seed
+    for (const auto& cluster : clusters) {
+      if(cluster.empty()) continue;
+      if(_debug > 1) printf("  Checking cluster of %zu tracks\n", cluster.size());
+      const KalSeedPtr* seedPtr = &(cluster.front());
```

Comment:
Pointers to ptrs is a convoluted data type I suggest to avoid. Since you eventually need a deep copy Ptr for your output collection, I suggest to start with that (ie push_back cluster.front() to your output collection), and then re-assign it as necessary. That simplifies the memory manipulation, and avoids the final translation.

Reply:
I was definitely not satisfied with this implementation, but I forgot to clean this up after implementing the matching module. I've cleaned this up with your suggestion, but I'm sure it could use more cleaning.